Fix for constant assignments to optional fields in Go.

davinchia opened the pull request:
As per ticket https://issues.apache.org/jira/browse/THRIFT-4253.

As I dug through the code I realised the problem was not only limited to optional strings, and that the generator was not properly generating optionals for all types.

davinchia commented:
https://stackoverflow.com/questions/30716354/how-do-i-do-a-literal-int64-in-go is a good summary of the different ways to assign constant values in Go.

I decided to use an anonymous struct since it seems to be the most self contained and efficient way to do so.

davinchia commented:
@Jens-G 
This is the first time I'm writing production C++ code so please feel free to tear this apart.

Jens-G commented on compiler/cpp/src/thrift/generate/t_go_generator.cc:
```diff
@@ -424,6 +424,52 @@ bool t_go_generator::is_pointer_field(t_field* tfield, bool in_container_value)
+    case t_base_type::TYPE_I8:
+      go_value_str += "int8}{";
+      go_value_str += std::to_string(static_cast<long long>(value->get_integer()));
+      break;
```
Three questions:
a) Why are these casts needed?
b) Can we move that whole method next to the other method? 
c) Why do we need two separate methods at all?

davinchia replied:
a) The casts are to avoid the `ambiguous calls` errors I was seeing on the integration tests.
b) c) I separated the methods as I had mini unit tests on my machine. I left it this way since I felt the main method was getting slightly too long. Combining both methods will remove the casting problem and I have no strong opinions so I will do that instead.

Jens-G commented:
Ping

davinchia commented:
@Jens-G Github must have hidden my comments. My latest commit combined the functions and removed the toString function as per your suggestions.

Not sure why one of the integration tests are failing. Logs show `/usr/include/php5/Zend/zend_hash.h:266:2:` is the cause, which is not something I modified.

jeking3 commented:
@davinchia we've cleared out the CI build issues, so if you could squash your changes to one commit then rebase on master, then do a force push to refresh this pull request, it'll kick off a new build.  Squashing will make it easier to apply the patch when the build completes successfully.

jeking3 commented:
@davinchia we've cleared out the CI build issues, so if you could squash your changes to one commit then rebase on master, then do a force push to refresh this pull request, it'll kick off a new build. Squashing will make it easier to apply the patch when the build completes successfully.

jeking3 commented:
@davinchia we've cleared out the CI build issues, so if you could squash your changes to one commit then rebase on master, then do a force push to refresh this pull request, it'll kick off a new build. Squashing will make it easier to apply the patch when the build completes successfully.

jeking3 commented:
Looks like the build was pretty much clean.  Any objections to a squash and merge?

davinchia commented:
No objections. Happy to squash and merge!

davinchia commented:
I don't have write access so I can't do this.

jeking3 commented:
No problem, now that we're on github as master I can squash and merge this with one click.

davinchia commented:
Thanks @jeking3! The only sad bit about this is since you merged this in, I don't get cred on the contributors page - dang!

jeking3 commented:
You are the commit author, that shows up in "git log".
